timemachine/cpp/src/gpu_utils.cuh

```diff
@@ -83,3 +84,81 @@ double __device__ __forceinline__ rmul_rn(double a, double b) { return __dmul_rn
 float __device__ __forceinline__ radd_rn(float a, float b) { return __fadd_rn(a, b); }
 
 double __device__ __forceinline__ radd_rn(double a, double b) { return __dadd_rn(a, b); }
+
+// If this were not int128s, we could do __shfl_down_sync, but only supports up to 64 values
+__device__ __forceinline__ void warp_reduce(volatile __int128 *shared_data, int thread_idx) {
+#pragma unroll 6
+    for (int i = 32; i > 0; i /= 2) {
+        if (thread_idx < i) {
+            shared_data[thread_idx] = shared_data[thread_idx] + shared_data[thread_idx + i];
+        }
+        __syncwarp();
+    }
+}
+
+// Taken originally from https://developer.download.nvidia.com/assets/cuda/files/reduction.pdf
+// but modified to account for post Cuda 9.0 which no longer implicitly syncs warps and the
+// fact that we don't need to launch recursive blocks as we can simply reduce the number of energies
+// that are accumulated.
+template <unsigned int BLOCK_SIZE>
+void __global__ k_accumulate_energy(
+    int N,
+    const __int128 *__restrict__ input_buffer, // [N]
+    __int128 *__restrict u_buffer              // [1]
+) {
+    __shared__ __int128 shared_mem[BLOCK_SIZE];
+    unsigned int tid = threadIdx.x;
+    if (tid >= BLOCK_SIZE) {
+        return;
+    }
+
+    unsigned int i = tid;
+    const unsigned int stride = BLOCK_SIZE * 2;
+    shared_mem[tid] = 0;
+    while (i < N) {
+        // Only add two values if the input buffer has enough input
+        shared_mem[tid] += input_buffer[i] + (i + BLOCK_SIZE < N ? input_buffer[i + BLOCK_SIZE] : 0);
+        i += stride;
+    }
+    //
+    __syncthreads();
+
+    // For larger blocks larger than a warp
```

Comment:
1) Since `BLOCK_SIZE` is a compile time constant, can we `static_assert` that `BLOCK SIZE <= 1024?`
2) Does this code break if `BLOCK_SIZE` is not a power of 2?

Reply:
Added static assertions. You are correct, need to be powers of two. BLOCK_SIZE=512 seems to be the best, but added support for 1024 in f335db89264349c37d13f05af03223c5b033bef4